File: cascadeflow/schema/domain_config.py
"""
Domain-Specific Cascade Configuration

This module provides the DomainConfig class for configuring domain-specific
cascade behavior. Each domain (CODE, MEDICAL, GENERAL, etc.) can have its own
drafter/verifier models, quality thresholds, and generation parameters.

Example:
    >>> from cascadeflow import CascadeAgent, DomainConfig
    >>>
    >>> code_config = DomainConfig(
    ...     drafter="deepseek-coder",
    ...     verifier="gpt-4o",
    ...     threshold=0.85,
    ...     temperature=0.2,
    ...     validation_method="syntax",
    ... )
    >>>
    >>> agent = CascadeAgent(
    ...     domain_configs={
    ...         "code": code_config,  # Use string keys
    ...     },
    ... )

Note:
    This module uses string domain identifiers to avoid circular imports.
    Domain strings match the Domain enum values in cascadeflow.routing.domain.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional, Union

# Domain string constants (matches routing.domain.Domain values)
# Using strings avoids circular imports with routing module
DOMAIN_CODE = "code"
DOMAIN_DATA = "data"
DOMAIN_STRUCTURED = "structured"
DOMAIN_RAG = "rag"
DOMAIN_CONVERSATION = "conversation"
DOMAIN_TOOL = "tool"
DOMAIN_CREATIVE = "creative"
DOMAIN_SUMMARY = "summary"
DOMAIN_TRANSLATION = "translation"
DOMAIN_MATH = "math"
DOMAIN_SCIENCE = "science"
DOMAIN_MEDICAL = "medical"
DOMAIN_LEGAL = "legal"
DOMAIN_FINANCIAL = "financial"
DOMAIN_GENERAL = "general"


class DomainValidationMethod(str, Enum):
    """Validation methods for domain-specific validation."""

    NONE = "none"
    SYNTAX = "syntax"  # Code/JSON syntax validation
    FACT = "fact"  # Fact-checking (medical, legal)
    SAFETY = "safety"  # Safety/toxicity checking
    QUALITY = "quality"  # General quality validation
    SEMANTIC = "semantic"  # ML-based semantic similarity
    CUSTOM = "custom"  # Custom validation function


@dataclass
class DomainConfig:
    """
    Domain-specific cascade configuration.

    Allows fine-grained control over how cascading works for each domain:
    - Model selection (drafter/verifier)
    - Quality thresholds
    - Generation parameters
    - Fallback behavior

    Attributes:
        drafter: Drafter model name or ModelConfig (cheaper, faster model)
        verifier: Verifier model name or ModelConfig (more capable model)
        threshold: Quality threshold (0-1) for accepting drafter responses
        validation_method: Validation method for this domain
        temperature: Temperature for generation (0-2)
        max_tokens: Maximum tokens to generate
        fallback_models: Fallback models to try if both fail
        require_verifier: Always use verifier, even if drafter passes
        adaptive_threshold: Enable adaptive threshold learning
        skip_on_simple: Skip verifier for trivial/simple queries
        enabled: Whether this domain config is enabled
        description: Human-readable description
        metadata: Additional metadata for custom use cases

    Example:
        >>> config = DomainConfig(
        ...     drafter="gpt-4o-mini",
        ...     verifier="gpt-4o",
        ...     threshold=0.85,
        ...     temperature=0.3,
        ... )
    """

    # Required: Model selection
    drafter: Union[str, Any]  # str or ModelConfig
    verifier: Union[str, Any]  # str or ModelConfig

    # Quality control
    threshold: float = 0.70
    validation_method: Union[str, DomainValidationMethod] = DomainValidationMethod.QUALITY

    # Generation parameters
    temperature: float = 0.7
    max_tokens: int = 1000

    # Fallback chain
    fallback_models: list[str] = field(default_factory=list)

    # Behavior flags
    require_verifier: bool = False
    adaptive_threshold: bool = True
    skip_on_simple: bool = True

    # Metadata
    enabled: bool = True
    description: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        """Validate configuration."""
        if not self.drafter:
            raise ValueError("DomainConfig: drafter is required")
        if not self.verifier:
            raise ValueError("DomainConfig: verifier is required")

        if not 0 <= self.threshold <= 1:
            raise ValueError(
                f"DomainConfig: threshold must be between 0 and 1, got {self.threshold}"
            )

        if not 0 <= self.temperature <= 2:
            raise ValueError(
                f"DomainConfig: temperature must be between 0 and 2, got {self.temperature}"
            )

        if self.max_tokens <= 0:
            raise ValueError(
                f"DomainConfig: max_tokens must be positive, got {self.max_tokens}"
            )

        # Convert string validation method to enum
        if isinstance(self.validation_method, str):
            self.validation_method = DomainValidationMethod(self.validation_method)

    def resolve_models(self, registry: "ModelRegistry") -> tuple[Any, Any]:
        """
        Resolve model names to ModelConfig objects.

        Args:
            registry: ModelRegistry instance

        Returns:
            Tuple of (drafter_config, verifier_config)
        """
        drafter = registry.get(self.drafter) if isinstance(self.drafter, str) else self.drafter
        verifier = registry.get(self.verifier) if isinstance(self.verifier, str) else self.verifier
        return drafter, verifier


# Built-in domain configurations (using string keys to avoid circular imports)
BUILTIN_DOMAIN_CONFIGS: dict[str, DomainConfig] = {
    DOMAIN_CODE: DomainConfig(
        drafter="deepseek-coder",
        verifier="gpt-4o",
        threshold=0.85,
        validation_method=DomainValidationMethod.SYNTAX,
        temperature=0.2,
        description="Optimized for code generation with syntax validation",
    ),
    DOMAIN_MEDICAL: DomainConfig(
        drafter="gpt-4o-mini",
        verifier="gpt-4",
        threshold=0.95,
        validation_method=DomainValidationMethod.FACT,
        temperature=0.1,
        require_verifier=True,
        description="High-accuracy medical responses with mandatory verification",
    ),
    DOMAIN_LEGAL: DomainConfig(
        drafter="gpt-4o-mini",
        verifier="gpt-4o",
        threshold=0.90,
        validation_method=DomainValidationMethod.FACT,
        temperature=0.2,
        description="Legal domain with fact-checking",
    ),
    DOMAIN_DATA: DomainConfig(
        drafter="gpt-4o-mini",
        verifier="gpt-4o",
        threshold=0.80,
        validation_method=DomainValidationMethod.SYNTAX,
        temperature=0.3,
        description="Data analysis and SQL with syntax validation",
    ),
    DOMAIN_MATH: DomainConfig(
        drafter="gpt-4o-mini",
        verifier="gpt-4o",
        threshold=0.90,
        validation_method=DomainValidationMethod.SYNTAX,
        temperature=0.1,
        description="Mathematical reasoning with high precision",
    ),
    DOMAIN_STRUCTURED: DomainConfig(
        drafter="gpt-4o-mini",
        verifier="gpt-4o",
        threshold=0.75,
        validation_method=DomainValidationMethod.SYNTAX,
        temperature=0.2,
        description="Structured data extraction (JSON/XML)",
    ),
    DOMAIN_GENERAL: DomainConfig(
        drafter="groq/llama-3.1-70b",
        verifier="gpt-4o",
        threshold=0.70,
        validation_method=DomainValidationMethod.QUALITY,
        temperature=0.7,
        description="Fast general-purpose queries",
    ),
}


def get_builtin_domain_config(domain: str) -> Optional[DomainConfig]:
    """
    Get a built-in domain configuration.

    Args:
        domain: Domain string (e.g., "code", "medical", "general")
                Can also be a Domain enum value (will use .value)

    Returns:
        DomainConfig if available, None otherwise

    Example:
        >>> config = get_builtin_domain_config("code")
        >>> print(config.drafter)  # "deepseek-coder"
    """
    # Handle both string and Domain enum
    domain_str = domain.value if hasattr(domain, 'value') else domain
    return BUILTIN_DOMAIN_CONFIGS.get(domain_str)


def create_domain_config(
    drafter: str,
    verifier: str,
    threshold: float = 0.70,
    validation_method: str = "quality",
    temperature: float = 0.7,
    **kwargs,
) -> DomainConfig:
    """
    Create a DomainConfig with convenience syntax.

    Args:
        drafter: Drafter model name
        verifier: Verifier model name
        threshold: Quality threshold (0-1)
        validation_method: Validation method name
        temperature: Generation temperature
        **kwargs: Additional DomainConfig parameters

    Returns:
        Configured DomainConfig instance

    Example:
        >>> config = create_domain_config(
        ...     drafter="gpt-4o-mini",
        ...     verifier="gpt-4o",
        ...     threshold=0.85,
        ... )
    """
    return DomainConfig(
        drafter=drafter,
        verifier=verifier,
        threshold=threshold,
        validation_method=validation_method,
        temperature=temperature,
        **kwargs,
    )

File: packages/core/src/config/domain-config.ts
/**
 * Domain-Specific Cascade Configuration
 *
 * This module provides the DomainConfig class for configuring domain-specific
 * cascade behavior. Each domain (CODE, MEDICAL, GENERAL, etc.) can have its own
 * drafter/verifier models, quality thresholds, and generation parameters.
 *
 * @example
 * ```typescript
 * import { DomainConfig, Domain } from '@cascadeflow/core';
 *
 * const codeConfig: DomainConfig = {
 *   drafter: 'deepseek-coder',
 *   verifier: 'gpt-4o',
 *   threshold: 0.85,
 *   temperature: 0.2,
 *   validationMethod: 'syntax',
 * };
 *
 * const agent = new CascadeAgent({
 *   domainConfigs: {
 *     [Domain.CODE]: codeConfig,
 *   },
 * });
 * ```
 */

import type { ModelConfig } from '../config';
import { Domain } from '../routers/domain-router';
import { ValidationMethod } from '../routers/cascade-pipeline';

/**
 * Validation method types for domain-specific validation.
 */
export type DomainValidationMethod =
  | 'none'
  | 'syntax'
  | 'fact'
  | 'safety'
  | 'quality'
  | 'semantic'
  | 'custom';

/**
 * Domain-specific cascade configuration.
 *
 * Allows fine-grained control over how cascading works for each domain:
 * - Model selection (drafter/verifier)
 * - Quality thresholds
 * - Generation parameters
 * - Fallback behavior
 */
export interface DomainConfig {
  /**
   * Drafter model - the cheaper, faster model that generates initial responses.
   * Can be a model name (string) or full ModelConfig.
   */
  drafter: string | ModelConfig;

  /**
   * Verifier model - the more capable model that validates/regenerates.
   * Can be a model name (string) or full ModelConfig.
   */
  verifier: string | ModelConfig;

  /**
   * Tool-specific drafter model (optional).
   * Falls back to drafter if not specified.
   * Used when tools are provided and domain-aware tool routing is enabled.
   */
  toolDrafter?: string | ModelConfig;

  /**
   * Tool-specific verifier model (optional).
   * Falls back to verifier if not specified.
   * Used when tools are provided and domain-aware tool routing is enabled.
   */
  toolVerifier?: string | ModelConfig;

  /**
   * Quality threshold (0-1) for accepting drafter responses.
   * Higher = stricter validation, more verifier usage.
   * @default 0.70
   */
  threshold?: number;

  /**
   * Validation method for this domain.
   * - 'syntax': Code/JSON syntax validation
   * - 'fact': Fact-checking (medical, legal)
   * - 'safety': Safety/toxicity checking
   * - 'quality': General quality validation
   * - 'semantic': ML-based semantic similarity
   * @default 'quality'
   */
  validationMethod?: DomainValidationMethod;

  /**
   * Temperature for generation (0-2).
   * Lower = more deterministic, higher = more creative.
   * @default 0.7
   */
  temperature?: number;

  /**
   * Maximum tokens to generate.
   * @default 1000
   */
  maxTokens?: number;

  /**
   * Fallback models to try if drafter and verifier both fail.
   */
  fallbackModels?: string[];

  /**
   * Always use verifier, even if drafter passes quality threshold.
   * Useful for high-stakes domains like medical.
   * @default false
   */
  requireVerifier?: boolean;

  /**
   * Enable adaptive threshold learning.
   * System learns optimal thresholds over time.
   * @default true
   */
  adaptiveThreshold?: boolean;

  /**
   * Skip verifier for trivial/simple queries.
   * Uses complexity detection to determine simplicity.
   * @default true
   */
  skipOnSimple?: boolean;

  /**
   * Whether this domain config is enabled.
   * @default true
   */
  enabled?: boolean;

  /**
   * Per-domain complexity handling.
   * Specifies which complexity levels should use cascade (try drafter first).
   * If undefined, defaults to all complexities using cascade.
   * @example ['trivial', 'simple', 'moderate', 'hard'] - EXPERT goes to verifier
   * @example ['trivial', 'simple', 'moderate', 'hard', 'expert'] - All cascade
   */
  cascadeComplexities?: Array<'trivial' | 'simple' | 'moderate' | 'hard' | 'expert'>;

  /**
   * Human-readable description of this configuration.
   */
  description?: string;

  /**
   * Additional metadata for custom use cases.
   */
  metadata?: Record<string, any>;
}

/**
 * Default domain configuration values.
 */
export const DEFAULT_DOMAIN_CONFIG: Required<Omit<DomainConfig, 'drafter' | 'verifier' | 'toolDrafter' | 'toolVerifier' | 'fallbackModels' | 'description' | 'metadata' | 'cascadeComplexities'>> = {
  threshold: 0.70,
  validationMethod: 'quality',
  temperature: 0.7,
  maxTokens: 1000,
  requireVerifier: false,
  adaptiveThreshold: true,
  skipOnSimple: true,
  enabled: true,
};

/**
 * Create a DomainConfig with default values filled in.
 */
export function createDomainConfig(config: DomainConfig): DomainConfig & { drafter: string | ModelConfig; verifier: string | ModelConfig } {
  return {
    drafter: config.drafter,
    verifier: config.verifier,
    toolDrafter: config.toolDrafter, // Optional, falls back to drafter
    toolVerifier: config.toolVerifier, // Optional, falls back to verifier
    threshold: config.threshold ?? DEFAULT_DOMAIN_CONFIG.threshold,
    validationMethod: config.validationMethod ?? DEFAULT_DOMAIN_CONFIG.validationMethod,
    temperature: config.temperature ?? DEFAULT_DOMAIN_CONFIG.temperature,
    maxTokens: config.maxTokens ?? DEFAULT_DOMAIN_CONFIG.maxTokens,
    fallbackModels: config.fallbackModels ?? [],
    requireVerifier: config.requireVerifier ?? DEFAULT_DOMAIN_CONFIG.requireVerifier,
    adaptiveThreshold: config.adaptiveThreshold ?? DEFAULT_DOMAIN_CONFIG.adaptiveThreshold,
    skipOnSimple: config.skipOnSimple ?? DEFAULT_DOMAIN_CONFIG.skipOnSimple,
    enabled: config.enabled ?? DEFAULT_DOMAIN_CONFIG.enabled,
    cascadeComplexities: config.cascadeComplexities, // Optional, can be undefined
    description: config.description ?? '',
    metadata: config.metadata ?? {},
  };
}

/**
 * Validate a DomainConfig and throw if invalid.
 */
export function validateDomainConfig(config: DomainConfig): void {
  if (!config.drafter) {
    throw new Error('DomainConfig: drafter is required');
  }
  if (!config.verifier) {
    throw new Error('DomainConfig: verifier is required');
  }

  if (config.threshold !== undefined) {
    if (config.threshold < 0 || config.threshold > 1) {
      throw new Error(`DomainConfig: threshold must be between 0 and 1, got ${config.threshold}`);
    }
  }

  if (config.temperature !== undefined) {
    if (config.temperature < 0 || config.temperature > 2) {
      throw new Error(`DomainConfig: temperature must be between 0 and 2, got ${config.temperature}`);
    }
  }

  if (config.maxTokens !== undefined) {
    if (config.maxTokens <= 0) {
      throw new Error(`DomainConfig: maxTokens must be positive, got ${config.maxTokens}`);
    }
  }
}

/**
 * Map of domain configs keyed by Domain.
 */
export type DomainConfigMap = Partial<Record<Domain, DomainConfig>>;

/**
 * Built-in domain configurations optimized for common use cases.
 * Updated December 2025 with GPT-5, Claude Opus/Sonnet/Haiku 4.5 models.
 */
export const BUILTIN_DOMAIN_CONFIGS: DomainConfigMap = {
  [Domain.CODE]: {
    drafter: 'deepseek-coder', // DeepSeek Coder - excellent code generation
    verifier: 'claude-opus-4-5-20251101', // Opus 4.5 - best code reasoning
    threshold: 0.85,
    validationMethod: 'syntax',
    temperature: 0.2,
    description: 'Code generation with DeepSeek drafter and Opus 4.5 verification',
  },
  [Domain.MEDICAL]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good medical knowledge ($0.25/M)
    verifier: 'claude-opus-4-5-20251101', // Opus 4.5 - best medical reasoning
    threshold: 0.95,
    validationMethod: 'fact',
    temperature: 0.1,
    requireVerifier: true,
    description: 'High-accuracy medical with mandatory Opus 4.5 verification',
  },
  [Domain.LEGAL]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good legal knowledge ($0.25/M)
    verifier: 'claude-opus-4-5-20251101', // Opus 4.5 - best legal reasoning
    threshold: 0.90,
    validationMethod: 'fact',
    temperature: 0.2,
    description: 'Legal domain with GPT-5 Mini and Opus 4.5 verification',
  },
  [Domain.FINANCIAL]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good financial ($0.25/M)
    verifier: 'gpt-5', // GPT-5 - excellent numerical reasoning
    threshold: 0.85,
    validationMethod: 'quality',
    temperature: 0.3,
    description: 'Financial analysis with GPT-5 Mini drafter and GPT-5 verifier',
  },
  [Domain.DATA]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good data analysis ($0.25/M)
    verifier: 'gpt-5', // GPT-5 - excellent data reasoning
    threshold: 0.80,
    validationMethod: 'syntax',
    temperature: 0.3,
    description: 'Data analysis with GPT-5 Mini drafter and GPT-5 syntax validation',
  },
  [Domain.MATH]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good math ($0.25/M)
    verifier: 'claude-opus-4-5-20251101', // Opus 4.5 - best mathematical reasoning
    threshold: 0.90,
    validationMethod: 'syntax',
    temperature: 0.1,
    description: 'Math with GPT-5 Mini drafter and Opus 4.5 verification',
  },
  [Domain.STRUCTURED]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good structured output ($0.25/M)
    verifier: 'gpt-5', // GPT-5 - excellent JSON/XML
    threshold: 0.75,
    validationMethod: 'syntax',
    temperature: 0.2,
    description: 'Structured extraction with GPT-5 Mini and syntax validation',
  },
  [Domain.CREATIVE]: {
    drafter: 'claude-3-5-haiku-20241022', // Claude Haiku - fast creative
    verifier: 'claude-sonnet-4-5-20250929', // Sonnet 4.5 - quality creative
    threshold: 0.60,
    validationMethod: 'quality',
    temperature: 0.9,
    description: 'Creative writing with Claude Haiku and Sonnet 4.5 verification',
  },
  [Domain.GENERAL]: {
    drafter: 'claude-3-5-haiku-20241022', // Claude Haiku - fast general ($0.25/$1.25/M)
    verifier: 'claude-sonnet-4-5-20250929', // Sonnet 4.5 - quality verification
    threshold: 0.70,
    validationMethod: 'quality',
    temperature: 0.7,
    description: 'General queries with Claude Haiku and Sonnet 4.5 verification',
  },
  [Domain.CONVERSATION]: {
    drafter: 'claude-3-5-haiku-20241022', // Claude Haiku - natural conversation
    verifier: 'gpt-5', // GPT-5 - excellent conversation
    threshold: 0.65,
    validationMethod: 'quality',
    temperature: 0.8,
    description: 'Conversational with Claude Haiku and GPT-5 verification',
  },
  [Domain.TOOL]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good tool calling ($0.25/M)
    verifier: 'gpt-5', // GPT-5 - excellent function calling
    threshold: 0.75,
    validationMethod: 'syntax',
    temperature: 0.2,
    description: 'Tool calling with GPT-5 Mini and GPT-5 verification',
  },
  [Domain.RAG]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good context handling ($0.25/M)
    verifier: 'claude-opus-4-5-20251101', // Opus 4.5 - best context synthesis
    threshold: 0.80,
    validationMethod: 'quality',
    temperature: 0.3,
    description: 'RAG with GPT-5 Mini and Opus 4.5 context verification',
  },
  [Domain.SUMMARY]: {
    drafter: 'claude-3-5-haiku-20241022', // Claude Haiku - fast summarization
    verifier: 'claude-sonnet-4-5-20250929', // Sonnet 4.5 - quality summaries
    threshold: 0.70,
    validationMethod: 'quality',
    temperature: 0.5,
    description: 'Summarization with Claude Haiku and Sonnet 4.5 verification',
  },
  [Domain.TRANSLATION]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - good multilingual ($0.25/M)
    verifier: 'gpt-5', // GPT-5 - excellent translation
    threshold: 0.80,
    validationMethod: 'quality',
    temperature: 0.3,
    description: 'Translation with GPT-5 Mini and GPT-5 verification',
  },
  [Domain.MULTIMODAL]: {
    drafter: 'gpt-5-mini', // GPT-5 Mini - vision capable ($0.25/M)
    verifier: 'claude-opus-4-5-20251101', // Opus 4.5 - best multimodal reasoning
    threshold: 0.75,
    validationMethod: 'quality',
    temperature: 0.4,
    description: 'Multimodal with GPT-5 Mini and Opus 4.5 verification',
  },
};

/**
 * Get a built-in domain configuration.
 */
export function getBuiltinDomainConfig(domain: Domain): DomainConfig | undefined {
  return BUILTIN_DOMAIN_CONFIGS[domain];
}

/**
 * Convert ValidationMethod enum to DomainValidationMethod string.
 */
export function validationMethodToDomain(method: ValidationMethod): DomainValidationMethod {
  switch (method) {
    case ValidationMethod.NONE:
      return 'none';
    case ValidationMethod.SYNTAX_CHECK:
      return 'syntax';
    case ValidationMethod.FACT_CHECK:
      return 'fact';
    case ValidationMethod.SAFETY_CHECK:
      return 'safety';
    case ValidationMethod.QUALITY_CHECK:
    case ValidationMethod.FULL_QUALITY:
      return 'quality';
    case ValidationMethod.SEMANTIC:
      return 'semantic';
    case ValidationMethod.CUSTOM:
      return 'custom';
    default:
      return 'quality';
  }
}

/**
 * Convert DomainValidationMethod string to ValidationMethod enum.
 */
export function domainValidationToMethod(validation: DomainValidationMethod): ValidationMethod {
  switch (validation) {
    case 'none':
      return ValidationMethod.NONE;
    case 'syntax':
      return ValidationMethod.SYNTAX_CHECK;
    case 'fact':
      return ValidationMethod.FACT_CHECK;
    case 'safety':
      return ValidationMethod.SAFETY_CHECK;
    case 'quality':
      return ValidationMethod.QUALITY_CHECK;
    case 'semantic':
      return ValidationMethod.SEMANTIC;
    case 'custom':
      return ValidationMethod.CUSTOM;
    default:
      return ValidationMethod.QUALITY_CHECK;
  }
}
